libs/animation/animation.ts

```diff
@@ -0,0 +1,136 @@
+                game.eventContext().registerFrameHandler(15, () => {
+                    _onAnimUpdate.forEach(element => {
+                        element();
+                    });
+                });
+            }
+            _onAnimUpdate.push(() => this.update());
+        }
+
+        update() {
+            let currentTime = control.millis();
+            let dt = currentTime - this.lastTime;
+            if (dt >= this.interval && this.frames.length) {
+                this.index = (this.index + 1) % this.frames.length;
+                this.lastTime = currentTime;
+            }
+        }
+
+        getImage() {
+            return this.frames[this.index];
+        }
+        
+        getAction() {
+            return this.action;
+        }
+
+        getInterval() {
+            return this.interval;
+        }
+
+        setInterval(interval: number) {
+            this.interval = interval;
+        }
+
+        //% blockId=addAnimationFrame
+        //% block="add frame $frame=screen_image_picker to $this=variables_get(anim)"
+        //% weight=40
+        addAnimationFrame(frame: Image) {
+            this.frames[++this.index] = frame;
+        }
+
+    }
+
+    //% shim=ENUM_GET
+    //% blockId=action_enum_shim
+    //% block="%arg"
+    //% enumName="ActionKind"
+    //% enumMemberName="action"
+    //% enumPromptHint="e.g. Walking, Idle, Jumping, ..."
+    //% enumInitialMembers="Walking, Idle, Jumping"
+    //% weight=10
+    export function _actionEnumShim(arg: number) {
+        // This function should do nothing, but must take in a single
+        // argument of type number and return a number value.
+        return arg;
+    }
+
+    /**
+     * Creates an animation
+     */
+    //% blockId=createAnimation 
+    //% block="create animation of $action=action_enum_shim with interval $interval ms"
+    //% interval.defl=1000
+    //% blockSetVariable="anim" 
+    //% weight=50
+    export function createAnimation(action: number, interval: number) {
+        return new Animation(action, interval);
+    }
+
+    /**
+     * attaches an animation to a sprite
+     */
+    //% blockId=attachAnimation 
+    //% block="attach animation $set=variables_get(anim) to sprite $sprite=variables_get(agent)"
+    //% weight=30
+    export function attachAnimation(sprite: Sprite, set: Animation) {
+        if (!_onSpriteUpdate) {
+            //First attach register the update call back.
+            //Priority 16 is slightly lower than 15 for animation update loop.
+            //This is allow the animation to complete, so we have the new display ready to go.
+            _onSpriteUpdate = [];
+            game.eventContext().registerFrameHandler(16, () => {
+                _onSpriteUpdate.forEach(element => {
+                    element();
+                });
+            });
+        }
+
+        _onSpriteUpdate.push(() => {
+            if (sprite._action === set.action) {
+                sprite.setImage(set.getImage())
```

Comment:
Better to not set the image every update because it will re-trigger the hitbox calculations